Phrase query evaluation fails due to 'unknown, ambiguous or inaccessible method phrase'

Original query: `(RUNTIME:[34 TO 52] or TEXT_TOKEN:"gte corp") (SCENE:fire)`

In this case, it seems that `Interpreter.call` is reacting to an argument mismatch when it tries to resolve a `content:phrase` method on the ContentFunctions class.

Remote debugging with Intellij, it appears that the first `phrase` method argument (i.e., local var `argv[0]`), which should be 'TEXT_TOKEN', is actually null at this point, resulting in the stacktrace below

Still investigating
```
[tserver.TabletServer] WARN : Failed to get multiscan result
java.util.concurrent.ExecutionException: java.lang.RuntimeException: Failure for query 028ccc46-bed5-4568-a65a-2afab26906d2
 .....
Caused by: org.apache.commons.jexl2.JexlException$Method: datawave.query.function.JexlEvaluation.<init>@48![1,57]:
 '(content:phrase(TEXT_TOKEN, termOffsetMap, 'gte', 'corp') 
&& TEXT_TOKEN == 'gte' 
&& TEXT_TOKEN == 'corp' 
&& (SCENE == 'fire')) 
|| (((_Value_ = true) 
  && ((_Bounded_ = true) 
  && (RUNTIME >= '+bE3.4' && RUNTIME <= '+bE5.2'))) 
  && (SCENE == 'fire'));' unknown, ambiguous or inaccessible method phrase
        at org.apache.commons.jexl2.Interpreter.call(Interpreter.java:1062)
        at org.apache.commons.jexl2.Interpreter.visit(Interpreter.java:1110)
        at datawave.query.jexl.DatawaveInterpreter.visit(DatawaveInterpreter.java:97)
        at org.apache.commons.jexl2.parser.ASTFunctionNode.jjtAccept(ASTFunctionNode.java:18)
        at org.apache.commons.jexl2.Interpreter.visit(Interpreter.java:1317)
        at datawave.query.jexl.DatawaveInterpreter.visit(DatawaveInterpreter.java:476)
        at org.apache.commons.jexl2.parser.ASTReference.jjtAccept(ASTReference.java:18)
        at datawave.query.jexl.DatawaveInterpreter.visit(DatawaveInterpreter.java:370)
        at org.apache.commons.jexl2.parser.ASTAndNode.jjtAccept(ASTAndNode.java:18)
        at datawave.query.jexl.DatawaveInterpreter.visit(DatawaveInterpreter.java:370)
        at org.apache.commons.jexl2.parser.ASTAndNode.jjtAccept(ASTAndNode.java:18)
        at datawave.query.jexl.DatawaveInterpreter.visit(DatawaveInterpreter.java:370)
        at org.apache.commons.jexl2.parser.ASTAndNode.jjtAccept(ASTAndNode.java:18)
        at datawave.query.jexl.DatawaveInterpreter.visit(DatawaveInterpreter.java:228)
        at org.apache.commons.jexl2.parser.ASTOrNode.jjtAccept(ASTOrNode.java:18)
        at org.apache.commons.jexl2.Interpreter.visit(Interpreter.java:946)
        at org.apache.commons.jexl2.parser.ASTJexlScript.jjtAccept(ASTJexlScript.java:38)
        at org.apache.commons.jexl2.Interpreter.interpret(Interpreter.java:232)
        at org.apache.commons.jexl2.ExpressionImpl.execute(ExpressionImpl.java:107)
        at datawave.query.function.JexlEvaluation.apply(JexlEvaluation.java:70)
        at datawave.query.function.JexlEvaluation.apply(JexlEvaluation.java:22)
        at datawave.query.iterator.QueryIterator$1.computeNext(QueryIterator.java:790)
        at datawave.query.iterator.QueryIterator$1.hasNext(QueryIterator.java:800)
        at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:42)
        at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:42)
        at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:42)
        at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:42)
        at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:42)
        at datawave.query.iterator.QueryIterator$1.computeNext(QueryIterator.java:788)
        at datawave.query.iterator.QueryIterator$1.hasNext(QueryIterator.java:800)
        at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:42)
        at datawave.query.iterator.pipeline.Pipeline.run(Pipeline.java:67)
        at datawave.query.iterator.pipeline.SerialIterator.startPipeline(SerialIterator.java:85)
        at datawave.query.iterator.QueryIterator.seek(QueryIterator.java:467)
        ... 11 more
```

--- a/warehouse/query-core/src/main/java/datawave/query/function/IndexOnlyContextCreator.java
+++ b/warehouse/query-core/src/main/java/datawave/query/function/IndexOnlyContextCreator.java
@@ -4,13 +4,13 @@ import com.google.common.base.Function;
 import com.google.common.collect.Multimap;
 import datawave.query.Constants;
 import datawave.query.attributes.Document;
+import datawave.query.collections.FunctionalSet;
 import datawave.query.composite.CompositeMetadata;
 import datawave.query.iterator.IndexOnlyFunctionIterator;
 import datawave.query.iterator.QueryOptions;
 import datawave.query.jexl.DatawaveJexlContext;
 import datawave.query.jexl.DelayedNonEventIndexContext;
 import datawave.query.jexl.IndexOnlyJexlContext;
-import datawave.query.jexl.visitors.DelayedNonEventSubTreeVisitor;
 import datawave.query.jexl.visitors.IteratorBuildingVisitor;
 import datawave.query.jexl.visitors.SetMembershipVisitor;
 import datawave.query.planner.DefaultQueryPlanner;
@@ -125,6 +125,7 @@ public class IndexOnlyContextCreator extends JexlContextCreator {
         final Tuple3<Key,Document,DatawaveJexlContext> tuple;
         if (null != from) {
             tuple = super.apply(from);
+            verifyContext(tuple.third());
         } else {
             tuple = null;
         }
@@ -197,6 +198,22 @@ public class IndexOnlyContextCreator extends JexlContextCreator {
         }
         
         return newContext;
+    }
+    
+    /**
+     * Ensures that downstream expectations are met with regard to the state of the given context
+     * 
+     * @param context
+     *            jexl context to verify, assumed to be non-null
+     */
+    private void verifyContext(DatawaveJexlContext context) {
+        // Make sure that index-only fields being queried are represented in the context, in
+        // order to avoid potential downstream eval misfire (see #1204)
+        for (String var : this.variables) {
+            if (this.indexOnlyFields.contains(var) && !context.has(var)) {
+                context.set(var, FunctionalSet.empty());
+            }
+        }
     }
     
     /*

--- a/warehouse/query-core/src/main/java/datawave/query/jexl/DelayedNonEventIndexContext.java
+++ b/warehouse/query-core/src/main/java/datawave/query/jexl/DelayedNonEventIndexContext.java
@@ -82,7 +82,7 @@ public class DelayedNonEventIndexContext extends DatawaveJexlContext {
         
         return delegate.get(name);
     }
-
+    
     /**
      * Use the IteratorBuildingVisitor limit to the current docRange to parse all delayed sub trees of the query. From those delayed sub trees initialize all
      * iterators matching the target field and aggregate all partial Documents into a list
@@ -106,7 +106,6 @@ public class DelayedNonEventIndexContext extends DatawaveJexlContext {
             
             // reset the root
             iteratorBuildingVisitor.resetRoot();
-
             
             // construct the index iterator for this node
             delayedNonEventNode.jjtAccept(iteratorBuildingVisitor, null);
@@ -155,27 +154,30 @@ public class DelayedNonEventIndexContext extends DatawaveJexlContext {
             }
         }
     }
-
+    
     @Override
     public boolean has(String name) {
         return delegate.has(name);
     }
-
+    
     @Override
     public void clear() {
         this.delegate.clear();
     }
-
+    
     @Override
     public int size() {
         return this.delegate.size();
     }
-
+    
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
-        if (!super.equals(o)) return false;
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        if (!super.equals(o))
+            return false;
         DelayedNonEventIndexContext that = (DelayedNonEventIndexContext) o;
         // @formatter:off
         return inclusive == that.inclusive
@@ -187,7 +189,7 @@ public class DelayedNonEventIndexContext extends DatawaveJexlContext {
                 && Objects.equal(fetched, that.fetched);
         // @formatter:on
     }
-
+    
     @Override
     public int hashCode() {
         // @formatter:off

--- a/warehouse/query-core/src/main/java/datawave/query/jexl/DelayedNonEventIndexContext.java
+++ b/warehouse/query-core/src/main/java/datawave/query/jexl/DelayedNonEventIndexContext.java
@@ -1,5 +1,6 @@
 package datawave.query.jexl;
 
+import com.google.common.base.Objects;
 import com.google.common.collect.Multimap;
 import datawave.query.attributes.Document;
 import datawave.query.attributes.ValueTuple;
@@ -81,7 +82,7 @@ public class DelayedNonEventIndexContext extends DatawaveJexlContext {
         
         return delegate.get(name);
     }
-    
+
     /**
      * Use the IteratorBuildingVisitor limit to the current docRange to parse all delayed sub trees of the query. From those delayed sub trees initialize all
      * iterators matching the target field and aggregate all partial Documents into a list
@@ -105,6 +106,7 @@ public class DelayedNonEventIndexContext extends DatawaveJexlContext {
             
             // reset the root
             iteratorBuildingVisitor.resetRoot();
+
             
             // construct the index iterator for this node
             delayedNonEventNode.jjtAccept(iteratorBuildingVisitor, null);
@@ -152,5 +154,52 @@ public class DelayedNonEventIndexContext extends DatawaveJexlContext {
                 }
             }
         }
+    }
+
+    @Override
+    public boolean has(String name) {
+        return delegate.has(name);
+    }
+
+    @Override
+    public void clear() {
+        this.delegate.clear();
+    }
+
+    @Override
+    public int size() {
+        return this.delegate.size();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        if (!super.equals(o)) return false;
+        DelayedNonEventIndexContext that = (DelayedNonEventIndexContext) o;
+        // @formatter:off
+        return inclusive == that.inclusive
+                && Objects.equal(delegate, that.delegate)
+                && Objects.equal(delayedNonEventFieldMap, that.delayedNonEventFieldMap)
+                && Objects.equal(docRange, that.docRange)
+                && Objects.equal(columnFamilies, that.columnFamilies)
+                && Objects.equal(equality, that.equality)
+                && Objects.equal(fetched, that.fetched);
+        // @formatter:on
+    }
+
+    @Override
+    public int hashCode() {
+        // @formatter:off
+        return Objects.hashCode(
+                super.hashCode(),
+                delegate,
+                delayedNonEventFieldMap,
+                docRange,
+                columnFamilies,
+                inclusive,
+                equality,
+                fetched);
+        // @formatter:on
     }
 }

